ext/opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py

```diff
@@ -0,0 +1,117 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-dbapi package allows tracing queries made by the
+ibraries following Ptyhon Database API specification:
+https://www.python.org/dev/peps/pep-0249/
+"""
+
+import typing
+
+import wrapt
+
+from opentelemetry.trace import SpanKind, Tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+QUERY_WRAP_METHODS = ["execute", "executemany", "callproc"]
+
+
+class DatabaseApiTracer:
+    def __init__(
+        self, tracer: Tracer, database_component: str, database_type="sql"
+    ):
+        if tracer is None:
+            raise ValueError("The tracer is not provided.")
+        self._tracer = tracer
+        self._database_component = database_component
+        self._database_type = database_type
+        self._connection_props = {}
+
+    def wrap_connect(
+        self,
+        wrapped: typing.Callable[..., any],
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch connect method to add tracing.
+        """
+        connection = wrapped(*args, **kwargs)
+        self._connection_props = {
+            "database": connection.database,
+            "port": connection.server_port,
+            "host": connection.server_host,
+            "user": connection.user,
+        }
+
+        wrapt.wrap_function_wrapper(connection, "cursor", self.wrap_cursor)
+        return connection
+
+    # pylint: disable=unused-argument
+    def wrap_cursor(
+        self,
+        wrapped: typing.Callable[..., any],
+        instance: typing.Any,
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch cursor instance in a specific connection.
+        """
+        cursor = wrapped(*args, **kwargs)
+        for func in QUERY_WRAP_METHODS:
+            wrapt.wrap_function_wrapper(cursor, func, self.add_span)
+        return cursor
+
+    # pylint: disable=unused-argument
+    def add_span(
+        self,
+        wrapped: typing.Callable[..., any],
+        instance: typing.Any,
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch execute, executeMany and callproc methods in cursor and create span.
+        """
+        name = self._database_component
+        database = self._connection_props.get("database", "")
+        if database:
+            name += "." + database
+        query = args[0] if args else ""
+        # Query with parameters
+        if len(args) > 1:
+            query += " params=" + str(args[1])
```

Comment:
I don't think we should add the params, this would make the query SQL invalid.

Reply:
I'm not updating the actual SQL query here, this variable is used to store db.statement field in the span, I can remove params from here to avoid sensitive information to be stored, but I feel some people would want to have full query information in the telemetry

Reply:
Maybe rename to `statement`?

Reply:
Yeah, I understand that the actual app will still work as before. But `db.statement` is supposed to contain the database statement, not any string that starts with a database statement. For example, a backend may want to parse the SQL query to find affected table names, etc. This becomes significantly harder when every integration appends some information in some nonstandard way.

Reply:
Added a new attribute called db.parameters to store these